react/features/base/session/middleware.js

```diff
@@ -0,0 +1,148 @@
+import { MiddlewareRegistry } from '../redux';
+import { toURLString } from '../util';
+
+import { createSession, endSession, sessionFailed, sessionStarted, sessionTerminated } from './actions';
+import { SESSION_CREATED, SESSION_FAILED, SESSION_STARTED, SESSION_TERMINATED } from './actionTypes';
+import { findSessionForConference, findSessionForConnection, findSessionForLocationURL } from './selectors';
+
+
+MiddlewareRegistry.register(({ dispatch, getState }) => next => action => {
+    const result = next(action);
+
+    switch (action.type) {
+    case CONFERENCE_FAILED: {
+        const { conference, error } = action;
+        const { recoverable } = error;
+        const session = findSessionForConference(getState(), conference);
+
+        if (session) {
+            session.conference = null;
+            if (typeof recoverable === 'undefined' || recoverable === false) {
+                session.conferenceFailed = true;
+                if (session.connection) {
+                    // The sessionFailed is expected to be dispatched from either CONNECTION_DISCONNECTED
+                    // or CONNECTION_FAILED handler in this middleware. The purpose is to delay the SESSION_FAILED until
+                    // the XMPP connection is properly disposed.
+                    dispatch(endSession(session));
+                } else {
+                    dispatch(sessionFailed(session));
+                }
+            }
+        } else {
+            console.error('CONFERENCE_FAILED - no session found');
+        }
+        break;
+    }
+    case CONFERENCE_LEFT: {
+        const { conference } = action;
+        const session = findSessionForConference(getState(), conference);
+
+        if (session) {
+            session.conference = null;
+
+            // If there's any existing connection wait for it to be closed first
+            session.connection || dispatch(sessionTerminated(session));
+        } else {
+            console.error('CONFERENCE_LEFT - no session found');
+        }
+        break;
+    }
+    case CONFERENCE_JOINED: {
+        const { conference } = action;
+        const session = findSessionForConference(getState(), conference);
+
+        if (session) {
+            dispatch(sessionStarted(session));
+        } else {
+            console.error('CONFERENCE_JOINED - no session found');
+        }
+
+        break;
+    }
+    case CONFERENCE_WILL_JOIN: {
+        const { conference } = action;
+        const { connection } = conference;
+
+        const session = findSessionForConnection(getState(), connection);
+
+        if (session) {
+            session.conference = conference;
+        } else {
+            console.error('CONFERENCE_WILL_JOIN - no session found');
+        }
+        break;
+    }
+    case CONFIG_WILL_LOAD: {
+        const { locationURL, room } = action;
+
+        room && room.length && dispatch(createSession(locationURL, room));
```

Comment:
Do you mean dispatch the load config from there ? So the thing is with the welcome page the config is loaded even fi there's no `room`, so this would not work. Maybe I don't understand what you mean exactly... Do you mean to have the session created, before the config is loaded ? How about dispatching `APP_NAVIGATE` and put the logic there which kills old sessions and creates a new one ?

Reply:
> Do you mean to have the session created, before the config is loaded ? How about dispatching `APP_NAVIGATE` and put the logic there which kills old sessions and creates a new one ?

Sounds like a plan, let's try that!

Reply:
> I'd say we should dispatch createSession first, and then dispatch the load from there 

So the problem is that we will not create a session if the config is being loaded for the welcome page. Do you mean to have all the logic of parsing uri and managing the load config flow in the `base/session` ? Then maybe the session could be created conditionally.